src/types.ts

```diff
@@ -10,11 +10,15 @@ export type IMethodName =
   | 'EmailSubmission/changes'
   | 'EmailSubmission/set';
 
+export type IErrorName = 'error';
+
 /**
  * See https://jmap.io/spec-core.html#the-invocation-data-type
- * [ name, arguments, id ]
+ * [ name, arguments, methodCallId ]
  */
-export type IMethodCall = [IMethodName, IArguments, string];
+export type IInvocation<ArgumentsType> =
+  | [IMethodName, ArgumentsType, string]
+  | [IErrorName, IError, string];
```

Comment:
Mmm I think we should get rid of this or and treat error as a regular Invocation.

Reply:
Maybe you're right, I am wondering why I did not write this library in Js with no typing at all, it would have been way simpler.

Reply:
Please don't be sarcastic.

IError is a ArgumentsType, 'error' is a response name, I would prefer having a IName type being either a IMethodName or a IError.

You see, easy, and you stick better to the spec <3.

By the way that string have a name https://jmap.io/spec-core.html#the-invocation-data-type ; `method call id`. 

(Giving specilized types rather than using primitive types is called strong typing and is definitly a good practice. From my (scala) experience time spent crafting a good type system is repayed 10 fold down the line.)

Reply:
I don't know all possible features of TypeScript, but as far as I know `|` is a [union](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types) operator, that tells TypeScript compiler, that variable/type can be any of mentioned types. The syntax is as follows `type1 | type2 | ...`. So it requires value on both sides of the pipe.

If I am reading this correctly, then here 
```ts
export type IInvocation<ArgumentsType> =
  | [IMethodName, ArgumentsType, string]
  | [IErrorName, IError, string];
```

it starts with pipe, not with a type.

Reply:
@jeserkin This syntax (added in TypeScript 2.1.4 https://github.com/microsoft/TypeScript/issues/12071) is easier to read (every line is at the same level).  Tools like prettier (formatter) will add it automatically on the first line if the type union is multi-line.

Reply:
> Please don't be sarcastic.

That was not sarcasm.

> IError is a ArgumentsType, 'error' is a response name, I would prefer having a IName type being either a IMethodName or a IError.

Ok, I made a `IInvocationName` type, is it ok?

> By the way that string have a name https://jmap.io/spec-core.html#the-invocation-data-type ; `method call id`.
> 
> (Giving specilized types rather than using primitive types is called strong typing and is definitly a good practice. From my (scala) experience time spent crafting a good type system is repayed 10 fold down the line.)

Ok, added name tuples to the `IInvocation` type.